### PR TITLE
[api][SIMS #28]Reassessment - Enable NOA view and approval

### PR DESCRIPTION
**Pull request by andrewsignori-aot**
Adjusted NOA endpoints to use the new assessment structure. These endpoints are still under the application controller but will be moved to the assessment controller in the future.

This will allow a generated assessment to be visualized and confirmed by the student avoiding changes in the UI for now.

DB columns will be removed in an upcoming PR.

### CODE REVIEW

File: sources/packages/api/src/route-controllers/application/application.controller.ts [hunk 1]
```diff
@@ -331,24 +334,34 @@ export class ApplicationController extends BaseController {
...
+    // TODO: temporary code to allow the assessment to be retrieved using the application id.
+    const [originalAssessment] =
+      await this.assessmentService.getAssessmentsByApplicationId(
         applicationId,
-        student.id,
+        AssessmentTriggerType.OriginalAssessment,
       );
-    if (!application) {
+    // TODO: end of the temporary code that will be removed.
+
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/application/application.controller.ts:345]: extra line

File: sources/packages/api/src/route-controllers/application/application.controller.ts [hunk 2]
```diff
@@ -394,14 +408,28 @@ export class ApplicationController extends BaseController {
...
+      await this.assessmentService.studentConfirmAssessment(
+        originalAssessment.id,
+        student.id,
+      );
+    } catch (error) {
+      switch (error.name) {
+        case ASSESSMENT_NOT_FOUND:
+          throw new NotFoundException(error.message);
+        case ASSESSMENT_INVALID_OPERATION_IN_THE_CURRENT_STATE:
+          throw new UnprocessableEntityException(error.message);
```
Comment (ann-aot) [inline sources/packages/api/src/route-controllers/application/application.controller.ts:429]: are we planning to add @apitags for these errors, or we are planning to do it all in a future ticket? just a qn

File: sources/packages/api/src/services/student-assessment/student-assessment.service.ts [hunk 3]
```diff
@@ -79,6 +79,79 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
...
+  ): Promise<StudentAssessment[]> {
+    const query = this.repo
+      .createQueryBuilder("assessment")
+      .select(["assessment.id"])
+      .innerJoin("assessment.application", "application")
+      .where("application.id = :applicationId", { applicationId });
+    if (triggerType) {
+      query.andWhere("assessment.triggerType = :triggerType", { triggerType });
+    }
+
```
Comment (ann-aot) [inline sources/packages/api/src/services/student-assessment/student-assessment.service.ts:151]: extra line

File: sources/packages/api/src/services/student-assessment/student-assessment.service.ts [hunk 4]
```diff
@@ -237,4 +310,49 @@ export class StudentAssessmentService extends RecordDataModelService<StudentAsse
       { noaApprovalStatus: status },
     );
   }
+
+  /**
+   * Updates assessment and application statuses when
+   * the student is confirming the NOA (Notice of Assessment).
+   * @param assessmentId assessment id to be updated.
+   * @param studentId student confirming the NOA.
+   * @returns update result. One and only one record
```
Comment (ann-aot) [inline sources/packages/api/src/services/student-assessment/student-assessment.service.ts:319]: `only one record`?

File: sources/packages/api/src/route-controllers/application/application.controller.ts (hunk 1 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/application/application.controller.ts:345]: This todo is to show what is supposed to be temporary. I am not sure if it should be considered an invalid extra line. Please see below what it looks like.

image

File: sources/packages/api/src/route-controllers/application/application.controller.ts (hunk 2 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/route-controllers/application/application.controller.ts:429]: This controller will be moved and changed so I avoided adding the tags here.

File: sources/packages/api/src/services/student-assessment/student-assessment.service.ts (hunk 3 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/student-assessment/student-assessment.service.ts]: Usually, we add some brack lines in the middle of the method body for the sake of making it clear. Here it is not the case that is why I am removing it, but in general it should not be a problem 😉

File: sources/packages/api/src/services/student-assessment/student-assessment.service.ts (hunk 4 above)
Comment (andrewsignori-aot) [reply sources/packages/api/src/services/student-assessment/student-assessment.service.ts]: Changed!

**Review by ann-aot — APPROVED** [review]
👍

**Review by dheepak-aot — APPROVED** [review]
Nice work @andrewsignori-aot 👍

**Review by henryoforeh-aot — APPROVED** [review]

**Review by guru-aot — APPROVED** [review]
LGTM, Nice work. Easy to follow

**Pull request merged**